### PR TITLE
added dsf support, reusing the existing ID3 parser

### PR DESCRIPTION
**Pull request by ilkomiliev**
please note that this is my very first golang work, so any feedback is appreciated. Still have no clue how to prepare tests, but already questions about this: 

1) are you including the necessary music files also in the repo (in case of dsf the files could be huge - the one I've tested with is 300 mb). 

2) copyrights: I think some dsd files are provided for test purposes on the web, but haven't check.

### CODE REVIEW

**Comment by dhowden** [comment]
Hey,  thanks for the contribution!

I'll try to have a look at the code in the next few days.

To answer your questions:  The tag data is in the first few KB of the file, so you shouldn't need to include much for a full test.  I don't have any DSF files myself, so if you could post a source of test files here that would be great.

**Comment by jooola** [comment]
@ilkomiliev If you have a sample file, you can add some testing.

Just check how the [tag_test.go](https://github.com/dhowden/tag/blob/master/tag_test.go) file works.

I couldn't find any DSF file, that's weird. So you could create on using the `testdata/with_tags/sample.flac` file.

**Comment by ilkomiliev** [comment]
should be ok now. Following site provides a lot of DSF files: 
[2L](http://www.2l.no/hires/index.html)
I've tested:
[this one](http://www.lindberg.no/hires/test/2L-056_03_stereo_DSD64.dsf)

File: dsf.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+type metadataDSF struct {
+	id3       Metadata
+	fileSize  int
```
[the remark this answers is not in the document]
Comment (ilkomiliev) [reply dsf.go]: done, the intention here was to provide also stream data, but let's keep it separately

File: dsf.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	chunkSize int
+}
+
+func (m *metadataDSF) readDSFMetadata(r io.ReadSeeker) error {
```
[the remark this answers is not in the document]
Comment (ilkomiliev) [reply dsf.go]: yeah - thanks for the help here!

File: dsf.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	return s
+}
+
+func hex2int(hexStr string) int {
+	result, _ := strconv.ParseInt(hexStr, 16, 64)
+	return int(result)
+}
+
+func lsb(n3 []byte) int {
```
[the remark this answers is not in the document]
Comment (ilkomiliev) [reply dsf.go]: I've ended here adding getIntLittleEndian in utils.go and the corresponding test in utils_test.go. Hope this is ok

**Comment by jooola** [comment]
I tried to create a file that we could keep in the repo: 
[sample.zip](https://github.com/dhowden/tag/files/2407311/sample.zip)

Could you please try it ?

**Comment by ilkomiliev** [comment]
> I tried to create a file that we could keep in the repo:
> [sample.zip](https://github.com/dhowden/tag/files/2407311/sample.zip)
> 
> Could you please try it ?

Metadata Format: ID3v2.3
File Type: DSF
 Title: Test Title
 Album: Test Album
 Artist: Test Artist
 Composer: Test Composer
 Genre: Jazz
 Year: 2000
 Track: 3 of 0
 Disc: 2 of 0
 Picture: <nil>
 Lyrics:


"TYER": "2000"
"TALB": "Test Album"
"TPE1": "Test Artist"
"TPE2": "Test AlbumArtist"
"TPOS": "2"
"TRCK": "3"
"COMM": &tag.Comm{Language:"eng", Description:"", Text:"Test Comment\x00"}
"TCOM": "Test Composer"
"TCON": "Jazz"
"TIT2": "Test Title"

let me check then how to integrate it into the test suite and I'll update the PR with the tests

**Comment by ilkomiliev** [comment]
ok, now looks ok - I have to edit the dsf file to add the ttl tracks. Now the test is passing. Please review it again, thanks.

**Comment by ilkomiliev** [comment]
Hi,

can this be merged or something is still missing?

**Comment by dhowden** [comment]
Hi there!  Apologies, I've been preoccupied with other stuff.  As I don't have any DSF files myself, I would have to read the spec to verify that the code is correct (and hence I've been putting it off somewhat!).

If anyone else has DSF files that they can run though that would be great, otherwise I hope to look at this at the weekend.

**Comment by jooola** [comment]
@ilkomiliev Did you just give up ? Sad...

**Comment by ilkomiliev** [comment]
no, I've messed up the branches somehow, trying to clean up :-(
let me check again

**Comment by ilkomiliev** [comment]
ok, back again - good that github is so sophisticated!

**Comment by ilkomiliev** [comment]
sorry for that - it seems I mess it totally, I hope I've fixed everything now. Quite late already here - if something goes wrong again, I'll check it tomorrow.

**Pull request merged**